[86bzw6xkg][date-picker] calendar as accessible grid

msereniti opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Motivation and Context

This PR makes our calendars of all date-p family to be represneted as grid instead of long list of buttons.

I fixed a11y of custom day example as well.

## How has this been tested?

Manually with VO. Needs testing with NDVA

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [x] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [x] Nice improve.

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] My code follows the code style of this project.
- [x] I have updated the documentation accordingly or it's not required.
- [x] Unit tests are not broken.
- [x] I have added changelog note to corresponding `CHANGELOG.md` file with planned publish date.
- [ ] I have added new unit tests on added of fixed functionality.

ilyabrower commented on semcore/date-picker/src/components/Calendar.jsx:
```diff
@@ -329,13 +343,15 @@ class CalendarDaysRoot extends CalendarAbstract {
-        <CalendarWeekDays locale={locale} />
-        <SGridDays onMouseLeave={this.handleMouseLeave}>
+      <SCalendar render={Box} {...this.getRootProps()}>
+        <CalendarWeekDays locale={locale} role='row' />
```
why do we need `row` here? Should we `walk` through the header of Calendar?

msereniti replied:
It's required to put row between grid and cell

j-mnizhek commented:
Have a list of questions and comments, but asked about the PR purpose in the DM first, because maybe some of them are irrelevant.

Here are some first comments and issues I've found:

1. In the DatePicker example, the combined aria-label is read incorrectly: From the label (for= attribute): "Date range picker Month range picker". The "Month range picker" needs to be removed. There might be an issue with the uniqueness of IDs on the page.

2. In the examples with presets, the focus order in the dialog is different. In simple examples like DateRangePicker, the focus starts with the preset buttons, while in the pickers with period comparison, it starts with the month switching buttons. Can we use a consistent focus order to avoid confusing users? For example, starting with the month switching buttons?

3. Can we use `columnheader` for the weekday cells? They should then be read together with the days.

4. Can we set `aria-current="date"` on the cell for the selected date if it’s the current date?

5. We should add a `visually-hidden` `span` with the full name of the week for the weekdays.

6. In the WeekPicker, for some reasons I can’t navigate through the calendar using the keyboard 😔

msereniti commented:
> In the DatePicker example, the combined aria-label is read incorrectly: From the label (for= attribute): "Date range picker Month range picker". The "Month range picker" needs to be removed. There might be an issue with the uniqueness of IDs on the page.

Fixed the id

> In the examples with presets, the focus order in the dialog is different. In simple examples like DateRangePicker, the focus starts with the preset buttons, while in the pickers with period comparison, it starts with the month switching buttons. Can we use a consistent focus order to avoid confusing users? For example, starting with the month switching buttons?

Fixed

> Can we use columnheader for the weekday cells? They should then be read together with the days.

Changed

> Can we set aria-current="date" on the cell for the selected date if it’s the current date?

Changed

> We should add a visually-hidden span with the full name of the week for the weekdays.

Added aria-label instaed of  visually-hidden span. I guess it's a better option because it will not affect select-&-copy content but will replace the shorten version. Also I've removed attr attribute because we were using it in a reversed way – it should contain shorted version but we were putting the longer one.

> In the WeekPicker, for some reasons I can’t navigate through the calendar using the keyboard 😔

Updated example, it was very 💩

j-mnizhek commented:
Back here! Found some issues:

1. Cannot submit the choosen date from the calendar grid by keyboard. I'm pressing Enter on the screencast below, while I'm on some date, but nothing happens.
![Sep-18-2024 11-47-02](https://github.com/user-attachments/assets/f53603c5-9489-4d24-9333-599a749be6c5)

2. MonthRangePicker reads months as Month 1. Shouldn't just the month be read in this case? It's a bit confusing at the moment.
<img width="872" alt="image" src="https://github.com/user-attachments/assets/ccc0e116-b240-4755-917e-5f5d3b241e55">

3. Btw, can we change aria-labels for Prev and Next buttons for different DatePicker types? What I'm talking about:
- For all of them they now are read as **Previous period** and **Next period**, but it looks like the following labels will be more convenient:
--- for DatePicker, DateRangePeaker, Week picker: **Previous month** and **Next month**
--- for MonthPeaker and MonthRangePeaker: **Previous year** and **Next year**

j-mnizhek reviewed (approved):
Looks like we have all the necessary changes and fixes 😎

sheila-semrush commented on website/docs/components/date-picker/date-a11y.md:
```diff
@@ -7,31 +7,86 @@ tabs: Design('date-picker'), A11y('date-a11y'), API('date-api'), Example('date-c
+| <nobr>`DatePicker.Trigger` > `input`</nobr> | `aria-haspopup="dialog"` and inherits `aria-controls="IDREF"` from its parent. | `aria-haspopup` indicates that the element triggers a dialog.                                                  |
+|                                             | `aria-expanded="false/true"`                                                   | Indicates whether the dialog is open or not.                                                                   |
+|                                             | `role="combobox"`                                                              | Identifies the element as a combobox—a text input combined with a list of options.                             |
+|                                             | `aria-describedby="IDREF"`                                                     | Provides the input with an accessible description by linking to a hint that explains the required data format. |
```
don't get this thing with the hint, in the basic example it refers to an empty span. Is it correct? How is it supposed to work?

j-mnizhek replied:
Yeah, for sure, it's strange. The date format "MM/DD/YYYY" should be moved from aria-label to that span

<img width="622" alt="image" src="https://github.com/user-attachments/assets/0be3c2d1-53a7-42f8-9062-a84b2b6ae91e">

j-mnizhek replied:
Guys, @ilyabrower @msereniti can we move date format to the span that should contain the date format?

j-mnizhek replied:
Michael fixed this ✅

sheila-semrush commented:
@j-mnizhek I updated the Keyboard support table and added a couple other small fixes. You can review my last commit and edit over it if you want. Everything looks fine, only have one question about the hint (another comment).
